### DISCUSSION TITLE
Watched value and formState.isValid changes in different render cycles

### DISCUSSION THREAD

**Discussion by val-o** [react-hook-form/react-hook-form#3627] [Q&A]
**Describe the bug**
Watched values are out of sync with formState.isValid. In the provided sandbox you can notice that there are 2 renders in which formState.isValid changes out of sync with watched values. I suspect it's due to `await` at https://github.com/react-hook-form/react-hook-form/blob/master/src/useForm.ts#L343

**To Reproduce**
type 123456
see in console (we have 2 prints with different isValid values) 

**Codesandbox link (Required)**
[![Edit serene-glitter-6eq4w](https://codesandbox.io/static/img/play-codesandbox.svg)](https://codesandbox.io/s/serene-glitter-6eq4w?fontsize=14&hidenavigation=1&theme=dark)

![image](https://user-images.githubusercontent.com/9393189/101490057-05c60400-398c-11eb-8673-9c99e6f8ff5e.png)

**Expected behavior**
Watched values and formState.isValid changes together in one render.

ver 6.8.1

**Comment by bluebill1049** [comment]
Hi @val-o This is actually expected behavior.

- first render is trigger with `watch`, because that input is been watched
- second render is trigger with validating which all validation is async in hook form.

**Comment by vertic4l** [comment]
@bluebill1049 hi, i got a similar / different issue:

```
const schema = yup.object().shape({
    message: yup.string().required(),
});

const ConnectChat = (props: IConnectChatProps) => {
    const [isSubmitting, setSubmitting] = useState(false);

    const { register, handleSubmit, reset, formState } = useForm<FormData>({
        mode: "onChange",
        reValidateMode: "onChange",
        resolver: yupResolver(schema),
    });

    const onSubmit = handleSubmit(async (values: FormData) => {
        setSubmitting(true);

        await props.sendMessage(values.message);

        setSubmitting(false);
        reset();
    });

    useEffect(() => {
        const container = document.getElementsByClassName(styles.messages)[0];
        if (!container) return;

        container.scrollTop = container.scrollHeight;
    });

    const handleOnKeyPress = (event: any) => {
        if (event.key === "Enter") {
            onSubmit(event);
        }
    };

    console.log("outer", formState.isValid);

    const renderContent = () => {
        console.log(formState.isValid);

        return (
            <>
                <div className={styles.messages}>{renderMessages()}</div>

                <form
                    onSubmit={onSubmit}
                    className={styles.form}
                    onKeyDown={handleOnKeyPress}
                >
                    <Textarea
                        placeholder="Nachricht hier eingeben..."
                        reference={register}
                        name="message"
                    />

                    <div className={styles.submit}>
                        <Button
                            type={ButtonTypes.submit}
                            isLoading={isSubmitting}
                            disabled={!formState.isValid || isSubmitting}
                        >
                            senden
                        </Button>
                    </div>
                </form>
            </>
        );
    };

    const renderMessages = () => {
        const messages = props.messages.map((message, index) => {
            if (message.message === undefined) {
                return null;
            }

            const messageContainerClassname = clsx(styles.messageContainer, {
                [styles.messageContainer__customer]: message.isCustomer,
            });

            const messageClassname = clsx(styles.message, {
                [styles.message__customer]: message.isCustomer,
            });

            return (
                <div
                    key={message.message + index}
                    className={messageContainerClassname}
                >
                    {!message.isCustomer && (
                        <div className={styles.author}>{message.author}</div>
                    )}
                    <div className={messageClassname}>{message.message}</div>
                </div>
            );
        });

        if (messages.length === 0) {
            return null;
        }

        return messages;
    };

    return (
        <div className={styles.connectChat}>
            <div className={styles.header}>
                <span className={styles.icon} onClick={props.onClose}>
                    <Icon icon={IconsEnum.windowMinimize} weight="solid" />
                </span>
            </div>

            {!props.hasApiError && props.isLoading ? (
                <LoadingAnimation />
            ) : (
                renderContent()
            )}

            {props.hasApiError && !props.isLoading && <ApiError />}
        </div>
    );
};

export default ConnectChat;
```

within renderContent() formState.isValid should be false, but it is true. When i insert console.log formState.isValid has the right value (false).

What do i miss here?

**Comment by bluebill1049** [comment]
@vertic4l we have proxy around formState to improve perf (for users who are not subscribed to those states):

https://react-hook-form.com/api#formState

**Comment by vertic4l** [comment]
@bluebill1049 ok, i think i've got it now:

```
    const isValid = formState.isValid;
    const renderContent = () => {
        return (
            <>
                <div className={styles.messages}>{renderMessages()}</div>

                <form
                    onSubmit={onSubmit}
                    className={styles.form}
                    onKeyDown={handleOnKeyPress}
                >
                    <Textarea
                        placeholder="Nachricht hier eingeben..."
                        reference={register}
                        name="message"
                    />

                    <div className={styles.submit}>
                        <Button
                            type={ButtonTypes.submit}
                            isLoading={isSubmitting}
                            disabled={!isValid || isSubmitting}
                        >
                            senden
                        </Button>
                    </div>
                </form>
            </>
        );
    };
```

**Comment by vertic4l** [comment]
@bluebill1049 And thank you for your work. react-hook-form is just amazing!

**Comment by bluebill1049** [comment]
> @bluebill1049 And thank you for your work. react-hook-form is just amazing!

a pleasure to hear that.

**Comment by val-o** [comment]
Thanks for the explanation, but didn't you think that brings an inconsistent state (there is a value which is not valid, but formState.isValid is true)? It might lead to bugs if users will do some logic like useEffect(, [value, isValid]) (which my case actually =) ) which will be executed with false positive. Maybe we could do something like `isValidating` flag? so peeps will be aware that usage of watched value is not safe yet

Let me explain it better by example:

```
useEffect(() => {
 if(formState.isValid) {
     // Here devs might think that phoneNumber will be valid, but it won't be in case of it was previously valid and there was a change that made it invalid. So we will get 2 render and first one will be false positive, because formState.isValid will be true
     findUserByPhone(phoneNumber);
}
}, [phoneNumber, formState.isValid])
```

**Comment by bluebill1049** [comment]
> Thanks for the explanation, but didn't you think that brings an inconsistent state (there is a value which is not valid, but formState.isValid is true)? It might lead to bugs if users will do some logic like useEffect(, [value, isValid]) (which my case actually =) ) which will be executed with false positive. Maybe we could do something like `isValidating` flag? so peeps will be aware that usage of watched value is not safe yet
> 
> Let me explain it better by example:
> 
> ```
> useEffect(() => {
>  if(formState.isValid) {
>      // Here devs might think that phoneNumber will be valid, but it won't be in case of it was previously valid and there was a change that made it invalid. So we will get 2 render and first one will be false positive, because formState.isValid will be true
>      findUserByPhone(phoneNumber);
> }
> }, [phoneNumber, formState.isValid])
> ```

I hear your concern, you can vote for the following feature request: 

https://github.com/react-hook-form/react-hook-form/issues/2842

This is not a bug for the current state of hook form tho. 

- `watch` is return what's in the form
- alidation is async by default at `useForm`. 

A potential solution is to isolate `watch` with `useWatch`, so keep it away from `isValid`formState.

**Comment by val-o** [comment]
Thanks!
Could you please advise how to react to changes properly in my example (when you need to search for a user on phone change)
The only solution I see right now is to validate phoneNumber manually on each change, maybe there is a better one?
```
useEffect(() => {
 if(validatePhone(phoneNumber)) {
    findUserByPhone(phoneNumber);
}
}, [phoneNumber])
```

**Comment by bluebill1049** [comment]
I will have a play and get back to you with a solution @val-o.

**Comment by bluebill1049** [comment]
what about this approach?

https://codesandbox.io/s/relaxed-shirley-kxbmx?file=/src/App.tsx:536-537

**Comment by val-o** [comment]
That was my initial approach, but there is a case with copy paste - value changed but formState is still valid
or in case of phone number there might be a contact picker, which will replace value completely and it will be still valid, so side effect will note be executed

**Comment by bluebill1049** [comment]
> That was my initial approach, but there is a case with copy paste - value changed but formState is still valid
> or in case of phone number there might be a contact picker, which will replace value completely and it will be still valid, so side effect will note be executed

can you share that behavior in my CSB above?

**Comment by val-o** [comment]
You can  type 12345 and then just paste 123456, console.log will not be executed
Here is a sandbox https://codesandbox.io/s/hopeful-chaum-kt2d7?file=/src/App.tsx

Your sandbox is actually working, I think it's due to you pass formState as a dependency to useEffect, but then the effect function might be called as a reaction to other fields or formstate change, not necessarily 'name' field

**Comment by bluebill1049** [comment]
> Your sandbox is actually working, I think it's due to you pass formState as a dependency to useEffect, but then the effect function might be called as a reaction to other fields or formstate change, not necessarily 'name' field

I am not sure what do you referred to as `name` field, but `formState` has to be the dep in the useEffect.

**Comment by bluebill1049** [comment]
by the way, create a discussion for RFC `isValidating`

**Reply by bluebill1049** [reply]
https://github.com/react-hook-form/react-hook-form/discussions/3669
